server.py

```diff
@@ -1,102 +1,128 @@
-        NPN_helper.disable_token(token)
-        response = jsonify({CODE: 1,
-                            MSG: SUCCESS})
-    return response
-
-
-@app.route('/acknowledge_message_delivery', methods=[GET, POST])
-def update_last_hash():
-    last_hash = None
-    pubkey = None
-    expiration = None
-    response = jsonify({CODE: 0,
-                        MSG: PARA_MISSING})
-
-    if LASTHASH in request.args:
-        last_hash = request.args[LASTHASH]
-        if PUBKEY in request.args:
-            pubkey = request.args[PUBKEY]
-        if EXPIRATION in request.args:
-            expiration = request.args[EXPIRATION]
-
-    if request.json and LASTHASH in request.json:
-        last_hash = request.json[LASTHASH]
-        if PUBKEY in request.json:
-            pubkey = request.json[PUBKEY]
-        if EXPIRATION in request.json:
-            expiration = request.json[EXPIRATION]
-
-    if request.form and LASTHASH in request.form:
-        last_hash = request.form[LASTHASH]
-        if PUBKEY in request.form:
-            pubkey = request.form[PUBKEY]
-        if EXPIRATION in request.form:
-            expiration = request.form[EXPIRATION]
-
-    if last_hash and pubkey and expiration:
-        NPN_helper.update_last_hash(pubkey, last_hash, expiration)
-        response = jsonify({CODE: 1,
-                            MSG: SUCCESS})
-
-    return response
+PN_helper = PushNotificationHelper(logger)
+
+
+def register(args):
+    device_token = None
+    session_id = None
+    if TOKEN in args:
+        device_token = args[TOKEN]
+    if PUBKEY in args:
+        session_id = args[PUBKEY]
+
+    if device_token and session_id:
+        PN_helper.register(device_token, session_id)
+    elif device_token:
+        PN_helper.unregister(device_token)
+    else:
+        raise Exception(PARA_MISSING)
+
+
+def subscribe_to_closed_group(args):
+    closed_group_id = None
+    session_id = None
+    if PUBKEY in args:
+        session_id = args[PUBKEY]
+    if CLOSED_GROUP in args:
+        closed_group_id = args[CLOSED_GROUP]
+
+    if closed_group_id and session_id:
+        PN_helper.subscribe_closed_group(closed_group_id, session_id)
+    else:
+        raise Exception(PARA_MISSING)
+
+
+def unsubscribe_to_closed_group(args):
+    closed_group_id = None
+    session_id = None
+    if PUBKEY in args:
+        session_id = request.args[PUBKEY]
+    if CLOSED_GROUP in request.args:
+        closed_group_id = args[CLOSED_GROUP]
+
+    if closed_group_id and session_id:
+        PN_helper.unsubscribe_closed_group(closed_group_id, session_id)
+    else:
+        raise Exception(PARA_MISSING)
+
+
+def notify(args):
+    session_id = None
+    data = None
+    if SEND_TO in args:
+        session_id = args[SEND_TO]
+    if DATA in args:
+        data = args[DATA]
+
+    if session_id and data:
+        PN_helper.add_message_to_queue(args)
+    else:
+        raise Exception(PARA_MISSING)
+
+
+Routing = {'register': register,
```

Comment:
What about `unregister`?

Reply:
Oh, I see for that you're supposed to hit `register` with only a device token and no Session ID. It'd be a bit cleaner to have an explicit `unregister` endpoint :)